### PR TITLE
[TorchToTosa] Support scalar (rank-0) multi dim reductions

### PR DESCRIPTION
**Pull request by krotov01**
Pytorch allows reducing a rank-0 tensor with dim=0 or dim=-1, which yields the scalar itself. TOSA has no rank-0 axis to reduce over, so lower these cases as a no-op reduction over an empty set of axes.

### CODE REVIEW

**Review by Lallapallooza — CHANGES_REQUESTED** [review]
Please add e2e coverage.

**Comment by krotov01** [comment]
> Please add e2e coverage.

@Lallapallooza added, please check

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -1402,6 +1402,20 @@ class ConvertAtenMultipleDimsReductionOp
         reduceDims.push_back(i);
     }
 
+    // PyTorch accepts dim=0 and dim=-1 for scalar reductions. TOSA
+    // reduction ops currently require at least rank-1 tensors, and scalar
+    // reductions are semantically no-ops, so lower them as no-axis reductions.
+    if (inputRank == 0) {
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:1583]: Is this meant to cover all rank-0 reductions, or only scalar `sum.dim_IntList`? The title/body say rank-0 reductions, but one-dim reductions are still untouched.

```python
class M(torch.nn.Module):
    def forward(self, x):
        return torch.prod(x, dim=0)
```

This is valid PyTorch, exports to `aten.prod.dim_int`, and still fails TorchToTosa legalization. Same issue for scalar `prod(dim=-1)`, `any(dim=0/-1)`, and `all(dim=0/-1)`. If those are out of scope, can we narrow the PR title/body/tests to scalar `sum.dim_IntList`?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -1402,6 +1402,20 @@ class ConvertAtenMultipleDimsReductionOp
...
+      if (reduceDims.size() > 1)
+        return rewriter.notifyMatchFailure(
+            op, "scalar reduce dim appears multiple times");
+      if (!reduceDims.empty() && reduceDims.front() != 0 &&
+          reduceDims.front() != -1)
+        return rewriter.notifyMatchFailure(
+            op, "scalar reduce dim is statically invalid");
+      reduceDims.clear();
```
Comment (Lallapallooza) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:1591]: This turns some cases from a validation failure into invalid TOSA. Example:

```python
class M(torch.nn.Module):
    def forward(self, x):
        return torch.mean(x, dim=0, dtype=torch.float64)
```

Before this patch, rank-0 `dim=0` was rejected. Now it reaches the empty-axis path and lowers scalar f32 -> f64 as:

```mlir
tosa.identity %0 : (tensor<f32>) -> tensor<f64>
```

`tosa.identity` must be same-type. Scalar int32 `sum(dim=0)` has the same issue with `tensor<i32> -> tensor<i64>`. Can we keep the no-axis identity same-typed and cast explicitly when the result dtype changes?

File: test/Conversion/TorchToTosa/basic.mlir [hunk 3]
```diff
@@ -325,6 +325,44 @@ func.func @test_reduce_sum_dims$basic(%arg0: !torch.vtensor<[3,4,5,6],f32>) -> !
...
+
+// CHECK-LABEL:   func.func @test_reduce_sum_scalar_negative_dim$basic(
+// CHECK-SAME:                                                          %[[INPUT:.*]]: !torch.vtensor<[],f32>) -> !torch.vtensor<[],f32> {
+// CHECK:           %[[INPUT_TENSOR:.*]] = torch_c.to_builtin_tensor %[[INPUT]] : !torch.vtensor<[],f32> -> tensor<f32>
+// CHECK-NOT:       tosa.reduce_sum
+// CHECK:           %[[IDENTITY:.*]] = tosa.identity %[[INPUT_TENSOR]] : (tensor<f32>) -> tensor<f32>
+// CHECK:           %[[RESULT:.*]] = torch_c.from_builtin_tensor %[[IDENTITY]] : tensor<f32> -> !torch.vtensor<[],f32>
+// CHECK:           return %[[RESULT]] : !torch.vtensor<[],f32>
+// CHECK:         }
+func.func @test_reduce_sum_scalar_negative_dim$basic(%arg0: !torch.vtensor<[],f32>) -> !torch.vtensor<[],f32> {
```
Comment (Lallapallooza) [inline test/Conversion/TorchToTosa/basic.mlir:399]: The new tests only cover scalar f32 `sum(dim=0/-1)`.

If this PR intends to support rank-0 reductions generally, we should also cover other reductions that hit these templates: e.g. scalar `prod`/`any`/`all`, scalar `mean(dtype=...)`, integer scalar `sum`, and one non-sum list-dim user such as `torch.linalg.vector_norm(..., dim=0)`, maybe something else I might miss. If the intended scope is only scalar f32 `sum.dim_IntList`, can we narrow the PR title/body/tests to say that?

**Review by Lallapallooza — CHANGES_REQUESTED** [review]
Thanks for the changes, few comments.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (krotov01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:1583]: You're right, this PR only targets sum.dim_IntList. I'll rephrase it.

File: test/Conversion/TorchToTosa/basic.mlir (hunk 3 above)
Comment (krotov01) [reply test/Conversion/TorchToTosa/basic.mlir:399]: This change affects multiple dimensions reductions, so cannot be narrowed to only sum.dim_IntList. I added more tests (for vector_norm and mean) to cover the rest of the ops that use this code path, and leaving one dim reductions outside of the scope of that PR.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (krotov01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:1583]: Narrowed the title to multi dim reductions. More tests are added for other multiple dim reduction operators.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (krotov01) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:1591]: As discussed in DMs, this behaviour is not reproducible in e2e tests, cast is getting materialised correctly.

**Review by Lallapallooza — APPROVED** [review]
LGTM
